### ISSUE TITLE
Add handling of public inputs

### ISSUE THREAD

**Issue by grjte** [0xPolygonMiden/air-dsl#10]
We need to add a way to specify boundary constraints that rely on public inputs.

- [x] clarify syntax for declaring & referring to public inputs in the Air DSL
- [x] handle parsing of public input declarations and references
- [x] add public inputs to IR
- [x] handle codegen of public inputs and boundary constraints that depend on public inputs

**Comment by grjte** [comment]
@bobbinth I think we should keep the initial implementation of public inputs as simple as possible and add complexity iteratively, so I'm inclined towards fixed size vectors as [described here](https://github.com/maticnetwork/miden/discussions/254#discussioncomment-3444589). We can revisit things after we've added the other elements needed for v0.1 (periodic columns, auxiliary trace constraints, and random values).

To address the 2 points related to public inputs there:
> We may not know the size of public inputs ahead of time. This would mean we'd need to add support for vectors of unknown size.

I think we should revisit this later.

> If we need a public input which is just a single value, what's the best way to denote its type? Or should we assume that everything is a vector and something like foo: [1] is fine?

I think we should start by treating it as a vector and using [1] and then revisit later.

**Comment by bobbinth** [comment]
Yes, I agree with the approach.

**Comment by grjte** [comment]
Great! In that case, the initial declaration format for public inputs will be this:

```
pub_inputs:
  program_hash: [4]
  stack_inputs: [16]
  stack_outputs: [16]
```

**Comment by grjte** [comment]
Closed by #39
